rfc/rfc-88/rfc-88.md

```diff
@@ -0,0 +1,596 @@
+            this.name = name;
+        }
+        
+        public Builder withId(int id) {..}
+        ...
+        public NestedField build() {
+          return new Field(isOptional, id, name, type, doc, defaultValue);
+        }
+    }
+}
+```
+For the first stated goal, the following changes are proposed:
+* Substitute avro schema with `Schema` in commit metadata.
+  - Persistence: SCHEMA_KEY: Avro schema string -> `Schema` string
+  - Resolving compatibility: firstly try resolving as `Schema`, and fallback to Avro schema for compatibility.
+* Substitute Avro schema with `Schema` in reader/writer path of all engines.
+  - Replace schemas in configuration.
+  - Replace schemas in reader/writer building process.
+* Substitute Avro schema with `Schema` in log block.
+
+### Abstraction of Writers/Readers
+In the lake architecture, abstracting low-level file readers and writers is a crucial design choice. With a well defined 
+reader/writer abstraction, we can make the system more flexible, extensible, and reduce the efforts required for development
+and maintenance. Furthermore, the reading/writing behaviors among different engines can be unified, as well as sharing the performance optimization.
+#### Writer Abstraction
+![writer_abstract](writer_abstract.png)
+The proposed abstraction of writer  is shown in the diagram. It's generally divided into two levels:
+* FileWriter: FileWriter is the basic interface of writers for files in different formats. It's composed of a StructWriter and related writing interface provided by the underlying formats, such as ParquetWriter  and WriteSupport  for Parquet format.
+    ```java
+    public interface FileWriter<R> extends Closeable {
+        void write(R value);
+    }
+    ```
+* StructWriter: StructWriter is the abstraction for writing engine specific rows.
+    ```java
+    /**
+     * Base value writer interface for writing values of different types.
+     */
+    public interface ValueWriter<T> {
+      void write(T value);
+    }
+  
+    /**
+     * A base struct writer for writing struct record.
+     */ 
+    public abstract class StructWriter<R> implements ValueWriter<R> {
+       // writers for specific fields
+       private final ValueWriter<Object>[] writers;
+       
+       // get field from engine specific row
+       public abstract Object get(R row, int index);
+       
+       // write row
+       public abstract void write(R row);
+    }
+    ```
+With the above Writer abstraction, implementations of different file formats can be more consistent. Here is a pseudo 
+code to show how to build the writer for a specific format and engines.
+
+* Pseudo code for a specific format, assuming XFormat:
+```java
+/**
+ * Iterface for XFormat to provide wirte handle for writers.
+ */
+public interface XFormatWriteHandle {
+   // Actual write handler provided by the underlying format,
+   // e.g., RecordConsumer for Parquet.
+   void setWriteHandle(XXWriteHandle writeHandle);
```

Comment:
It may be a good opportunity to make the write handle layer aware of `HoodieStorage` so we have finer control over the IO as well